File: android/src/main/kotlin/tenko/rx_flutter_plugin/Logger.kt
package tenko.rx_flutter_plugin

import android.util.Log

object Logger {
    private val TAG = "rx_flutter_plugin"
    var debug = false

    fun d(msg: String) {
        if (debug) print(Log.DEBUG, msg)
    }

    fun d(t: Throwable? = null, msg: String) {
        if (debug) print(Log.DEBUG, msg, t)
    }

    fun v(msg: String) {
        if (debug) print(Log.VERBOSE, msg)
    }

    fun v(t: Throwable? = null, msg: String) {
        if (debug) print(Log.VERBOSE, msg, t)
    }

    fun e(msg: String) {
        if (debug) print(Log.ERROR, msg)
    }

    fun e(t: Throwable? = null, msg: String) {
        if (debug) print(Log.ERROR, msg, t)
    }

    fun w(msg: String) {
        if (debug) print(Log.WARN, msg)
    }

    fun w(t: Throwable? = null, msg: String) {
        if (debug) print(Log.WARN, msg, t)
    }

    fun i(msg: String) {
        if (debug) print(Log.DEBUG, msg)
    }

    fun i(t: Throwable? = null, msg: String) {
        if (debug) print(Log.DEBUG, msg, t)
    }

    private fun print(level: Int, msg: String, throwable: Throwable? = null) {
        if (throwable == null) {
            when (level) {
                Log.DEBUG -> Log.d(TAG, msg)
                Log.ERROR -> Log.e(TAG, msg)
                Log.INFO -> Log.i(TAG, msg)
                Log.WARN -> Log.w(TAG, msg)
                Log.VERBOSE -> Log.v(TAG, msg)
            }
        } else {
            when (level) {
                Log.DEBUG -> Log.d(TAG, msg, throwable)
                Log.ERROR -> Log.e(TAG, msg, throwable)
                Log.INFO -> Log.i(TAG, msg, throwable)
                Log.WARN -> Log.w(TAG, msg, throwable)
                Log.VERBOSE -> Log.v(TAG, msg, throwable)
            }
        }
    }
}

File: lib/logger.dart
class RxPluginLogger {
  static bool enabled = false;

  String tag;
  RxPluginLogger._();
  RxPluginLogger(String tag) {
    this.tag = tag;
  }

  void d(String message, [Exception exception, StackTrace stacktrace]) {
    if (enabled) {
      _printLog(_Level.debug, tag, message, exception, stacktrace);
    }
  }

  void w(String message, [Exception exception, StackTrace stacktrace]) {
    if (enabled) {
      _printLog(_Level.warning, tag, message, exception, stacktrace);
    }
  }

  void i(String message, [Exception exception, StackTrace stacktrace]) {
    if (enabled) {
      _printLog(_Level.info, tag, message, exception, stacktrace);
    }
  }

  void e(String message, [Exception exception, StackTrace stacktrace]) {
    if (enabled) {
      _printLog(_Level.error, tag, message, exception, stacktrace);
    }
  }

  static void _printLog(_Level level, String tag, String message, [Exception exception, StackTrace stacktrace]) {
    print("~RxFlutterPlugin~$tag/${_asString(level)}: $message");
    if (exception != null) {
      print("\tException: ${exception.toString()}");
      if (stacktrace != null) {
        print("\t\tStacktrace: ${stacktrace.toString()}");
      }
    }
  }

  static String _asString(_Level level) {
    switch (level) {
      case _Level.debug:
        return "D";
      case _Level.error:
        return "E";
      case _Level.info:
        return "I";
      case _Level.warning:
        return "W";
    }
  }
}

enum _Level {
  debug, warning, info, error
}